README.md
# shellswain

shellswain is a neighborly bash library you can use to build simpler event-driven bash profile/bashrc scripts and modules.

## What shellswain gives you

- loaded modules can emit their own arbitrary events and subscribe to events provided by other modules
- a core set of events emitted before the first prompt, before and after each command invocation, and before shell exit
- a way to set up command-specific before/run/after events (the run event enables you to customize how the command runs)
- shared metadata about the most-recent command (the command, when it ran, how long it took, what the exit status was, etc.)
- traps/signals automatically namespaced per script

These features form an efficient foundation for mutual cooperation between modules and user code. They make it easy to share access to scarce shell resources and minimize duplicated work.

## Incorporating shellswain

I package shellswain and its dependencies with Nix and resholve for my own use, so that's the easiest/recommended way to incorporate it into a project.

> **Note:** Aside from bash 5.1+, shellswain's dependencies are pure bash. It doesn't _require_ Nix and should be easy enough to package/vendor/inline outside of the Nix ecosystem. You'll also need:
> - signal/trap namespacing provided by https://github.com/abathur/comity
> - the event API provided by https://github.com/bashup/events (this is pulled in via comity, since comity also uses it)

You can find a real-world example of how I do this in https://github.com/abathur/shell-hag. That project is a little complex, so I'll break down the basic steps:

1. Include it in your Bash source. I use a guard to avoid wasting time sourcing it again it in case more than one module uses shellswain:

    ```bash
    if [[ -z "$SHELLSWAIN_ABOARD" ]]; then
        # shellcheck disable=SC1090
        source shellswain.bash
    fi
    ```

    For reference, here's [the equivalent statement in shell-hag](https://github.com/abathur/shell-hag/blob/c282e2b7581e57d5df2be261ffc505af0c92a091/hag.bash#L8-L12).

2. Package your script/module with Nix + [resholve](https://github.com/abathur/resholve) and supply shellswain as a dependency. Here's a basic skeleton:

    ```
    { lib
    , resholve
    , shellswain
    }:

    resholve.mkDerivation rec {
      pname = "your_project";
      version = "unreleased";

      src = lib.cleanSource ./.;
      # src = fetchFromGitHub {
      #   owner = "you";
      #   repo = "${pname}";
      #   rev = "v${version}";
      #   sha256 = "...";
      # };

      solutions = {
        profile = {
          scripts = [ "bin/your_module.bash" ];
          interpreter = "none";
          inputs = [ shellswain ];
        };
      };

      # ...
    }

    ```

    If it isn't clear how to turn this into a working Nix expression, I recommend referring to:
    - shellswain's own [shellswain.nix](shellswain.nix) is a simple, complete example of how to use resholve with Nix
    - resholve's Nix API is documented in the [nixpkgs README for resholve](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md)

## Using shellswain

> **Note:** If you just want to play around with shellswain, you can run `nix develop github:abathur/shellswain` to open a bare bash shell with shellswain pre-sourced. There are also some basic usage examples in the [examples](examples/) directory.

There are four main ~areas of shellswain's public API:

1. shellswain publishes events related to shell init/teardown:
    - `swain:before_first_prompt` (emitted the first time bash evaluates `PROMPT_COMMAND`)

      > **Note:** In early versions of shellswain this assumed ownership of `PROMPT_COMMAND`, though shellswain should now also be compatible with the new array-based `PROMPT_COMMAND`.

    - `swain:before_exit` (emitted on HUP and EXIT)

    Your code can subscribe to these events using the bashup.events API (see https://github.com/bashup/events for more). For example:

    ```bash
    event on swain:before_exit _your_teardown_function
    ```

2. shellswain publishes events before and after _every_ command invocation:
    - `swain:before_command` (emitted right after bash evaluates PS0)

      > **Note:** This does assume shellswain owns PS0. Your code can print whatever it would like as the before-command prompt by attaching a handler to the swain:before_command event.

    - `swain:after_command` (emitted each time bash evaluates `PROMPT_COMMAND` except the first; see `swain:before_first_prompt`)

3. if you instruct shellswain to "track" _specific_ commands, it will emit command-specific before/run/after events.

    Command-specific tracking is designed to encourage a lazy-initialization pattern. The goal is to avoid doing setup work for commands until the user actually invokes them. This may feel like a lot of conceptual overhead if you only want to track a single command--but the goal is making sure shell startup time is snappy even if users/modules are tracking scores of commands.

    - You call `swain.track <command> <your_init_callback>` to bootstrap deferred tracking for a command. shellswain will run `your_init_callback <command>` the first time the user invokes the tracked command.

        > **Note:** `swain.track` can only register one init callback, but other modules (or user code, if you're building a layer over shellswain) can use `swain.hook.init_command <command> <callback> <all other args>` to subscribe to a one-time event that shellswain will emit immediately after running the init callback.
        >
        > These callbacks are invoked as `callback <command> [<other args>...]`.

    - You can then use either kind of init callback to set up command-specific event listeners (and perform any other init you need).

        When users run a tracked command, shellswain emits three command-specific "phase" events: `before`, `run`, and `after`.

        You can set up a listener by calling:
        `swain.phase.listen <phase> <command> <callback> [<other args>...]`

        shellswain will invoke your callback as:
        `callback <command> [<other args from swain.phase.listen>...] [<args the user invoked command with>...]`

        > **Note:** shellswain also has a mechanism for currying additional arguments to a phase. You can call `swain.phase.curry_args <phase> <command> [<other args>...]` to inject args before those from the user's invocation.
        >
        > Instead of spending a second on a long computation in both the before and after phases, this enables you to compute it once in the before phase and curry the result to the after phase.

        The `run` phase is ~special--it's responsible for actually running the command. If none of your init callbacks register a run phase listener, shellswain will register a default runner (that just runs the command).

        If you register a run phase listener, make sure it runs the command!

4. shellswain maintains a global associative array, `swain`, with information about command run so that each plugin/module doesn't have to compute them independently. In the order they are recorded:

    Before the command runs:
    - `start_time`

       This is a human-readable datetime as reported by `printf '%(%a %b %d %Y %T)T'` (ex: "Sun Jan 15 2023 12:34:45").

    - `command_number`

       The command number as reported by `fc -lr -0`.

    - `command`

       The most-recently run command (unexpanded) as reported by `fc -lr -0`.

    - `start_timestamp`

       A microsecond-precision timestamp created by removing the `.` from `$EPOCHREALTIME` (ex: "1673807685512462").

       > Caution: This value will be slightly different during the `swain:before_command` and afterwards. It is recorded once before running any `swain:before_command` listeners, and updated after.
       >
       > shellswain does this so that it can both give _some_  timestamp to plugins that need one before the command runs and exclude the time `swain:before_command` listeners take to run from the `duration` it computes after the command runs.

    After the command runs (immediately before `swain:after_command` is emitted):
    - `end_timestamp`

       A microsecond-precision timestamp created by removing the `.` from `$EPOCHREALTIME` (ex: "1673800973404806").
    - `pipestatus`

       Command exit statuses as reported by `${PIPESTATUS[@]}` (ex: "0", "0 1 0").

    - `duration`

       How long the command took to run in microseconds as reported by `$((swain[end_timestamp] - swain[start_timestamp]))` (ex: "6687376").

       > Note: While the time taken to run `swain:before_command` and `swain:after_command` listeners are excluded from the calculated duration, it will still include whatever time it takes to run any phase listeners registered with `swain.phase.listen`.

    - `end_time`

       This is a human-readable datetime as reported by `printf '%(%a %b %d %Y %T)T'` (ex: "Sun Jan 15 2023 12:34:45").

    > **Caution:** shellswain updates the `swain` variable in place. If you use any after-command value during the `swain:before_command` event or any listener registered with swain.phase.listen, the values will still refer to the _previous_ command run.
    >
    > (This may sound like a footgun, but in some cases it is exactly the behavior you want. A session-oriented shell history plugin, for example, might need the previous command's end time to create a new history file when you go more than an hour without running a command.)



CHANGELOG.md
# Changelog

## v0.3.0 (Jan 21 2023)
- add and document some basic usage examples
- nix:
  - update comity
  - add devShell for trying out shellswain interactively

## v0.2.0 (Jan 17 2023)
- stop exporting PROMPT_COMMAND
  
  Not 100% certain on this change, but I see a prompt plugin maintainer with clear conviction that exporting it is wrong.

## v0.1.0 (Jan 15 2023)
- refactor to clearly define public/private API and regularize naming
- draft initial documentation for those APIs
- adapt to bash 5.1 (replace workarounds for 5.0 bugs)
- add tests

---

## Prerelease
In 2018, the code that would eventually become shellswain started off as part of a larger shell history module project (eventually named [shell-hag](https://github.com/abathur/shell-hag)) in my shell profile. 

In 2019 I realized that shell-hag's logistic layer might make a good foundation for other projects, so I teased it out and created shellswain.

Since they weren't _designed_ as separate projects, this process left some messy cross-cutting naming patterns and responsibility boundaries where the projects bled together. 

I didn't want to cut a release without fixing this, but figuring out how to better integrate Bash/Shell projects with Nix was higher up my priority list. That yak-shave, which led to https://github.com/abathur/resholve, has kept me from getting back to this for a while.

This is all to say that shellswain is newly released in 2023, but I've been using it daily via my Bash profile for about 4 years already.


tests/bare.bash
if [[ -n "$HAVE_SHELLSWAIN" ]]; then
	source "$SHELLSWAIN"
fi
set -x
for i in {1..100}; do
	ls
done


tests/helpers.bash
# setup_file(){
#   echo "test,shellswain?,comity?,duration" >> "$out/timings"
#   export TEST_TMP="$(mktemp -d)"
#   cp tests/*.{bats,bash} "$TEST_TMP"/ > /dev/null
#   pushd "$TEST_TMP"
#   PATH="$TEST_TMP:$PATH"
# } &> /dev/null

# teardown_file(){
#   popd
#   # separator
#   echo "" >> "$out/timings"
# }

setup() {
    {
        TEST_TMP="$(mktemp -d)"
        cp tests/*.{bats,bash} "$TEST_TMP"/ > /dev/null
        pushd "$TEST_TMP"
    } > /dev/null
}
teardown() {
    {
        popd > /dev/null
    } > /dev/null
}


tests/shellswain_track.bash
#!/usr/bin/env bash

fern(){
	echo hehe
}

alias alfred=fern

source shellswain.bash

swain.track "$1" _make_sure
ret=$?

eval "_make_sure(){
	echo 'captured call to $1 with args ${@:2}'
}
_test(){
	$@
}
"

_test

exit $ret


examples/inspect_command_results.bash
# register command-specific post-command actions
source shellswain.bash

__get_nix_build_result_before(){
	# check existing symlink and pass target path to the after phase
	swain.phase.curry_args "after" "nix-build" "$(readlink result)"
}

__compare_nix_build_result_after(){
	echo "coompare ${@@Q}"
	local after_result="$(readlink result)"
	if [[ "$1" != "$after_result" ]]; then
		# Note: won't duplicate entries if the callback is identical
		event on "nix-build:list-results" echo "'${swain[command]}' changed result to '$after_result'"
	fi
}

__track_nix_build_results(){
	swain.phase.listen "before" "$1" __get_nix_build_result_before
	swain.phase.listen "after" "$1" __compare_nix_build_result_after
}

nix-builds(){
	event emit "nix-build:list-results"
}

swain.track "nix-build" __track_nix_build_results


shellswain.bash
# shellcheck shell=bash

# Naming patterns:
# - shellswain-specific functions and events start with "swain"
# - namespace separators: '.' for functions and ':' for events
# - use function _names for likely-internal behavior

# shellcheck disable=SC2154
if [[ -n "$SHELLSWAIN_ABOARD" ]]; then
	return
fi

shopt -s expand_aliases

# set an array-safe prompt command
PROMPT_COMMAND=${PROMPT_COMMAND-}

# double-dip:
# - shellswain is loaded
# - record the index of shellswain's prompt command
SHELLSWAIN_ABOARD=${#PROMPT_COMMAND[@]}

# associative array, global, export
declare -Ax swain

# immediately record start time; if we do a good job of optimizing
# where this falls in the overall load it can report rough startup time
swain[start_timestamp]="${EPOCHREALTIME/.}"
# builtin printf much faster than external date
# shellcheck disable=SC2102,SC2183
printf -v swain[start_time] '%(%a %b %d %Y %T)T'
swain[command_number]=0
swain[command]="#shellswain init"

function swain._record_end()
{
	# record timestamp, which we use to calculate duration, first
	swain[end_timestamp]="${EPOCHREALTIME/.}"
	swain[pipestatus]=$*
	swain[duration]=$((swain[end_timestamp] - swain[start_timestamp]))

	# shellcheck disable=SC2102,SC2183
	printf -v swain[end_time] '%(%a %b %d %Y %T)T'
}

# save time if it's already loaded
[[ -v __comity_signal_map ]] || source comity.bash

# SHELLSWAIN CORE:
function swain._after_command() {
	swain._record_end "${PIPESTATUS[@]}"
	event emit "swain:after_command" "${PIPESTATUS[@]}"
	trap "swain._expand_PS0" SIGCHLD
}

function swain._before_first_prompt() {
	swain._record_end
	event emit "swain:before_first_prompt"
	PROMPT_COMMAND[$SHELLSWAIN_ABOARD]="swain._after_command"

	# The *INTENT* is that this only runs when no history could be loaded
	# However, up through at least bash 5.0, something like
	# echo $HISTCMD $(echo $HISTCMD $(set -H; echo $HISTCMD))
	# 254 1 1
	# demonstrates that history won't just "work" in the subshell, so
	# set -o history # we turned history on
	# AFAIK bash 5.1 has fixed this, so we'll disable the workaround.
	# Remove after 2023.

	# Just trying to trick bash into launching *exactly one* subshell
	# when it expands PS0 so we can trap the subshell exit. This enables
	# us to run code:
	#
	# - outside of the subshell context (to modify persistent variables)
	# - at the time bash expands PS0 (to track command start time)
	#
	# PS0 enables us to run arbitrary code before command time,
	# but it's all in a subshell (requiring files or hacks to send data
	# back to parent shell). SIGCHLD trap is the approach used by the
	# existing bash-preexec hook (which does give it at least one big
	# advantage: BASH_COMMAND is updated by the time debug trap runs,
	# but NOT by the time we trap this subshell exit; we can't use
	# BASH_COMMAND here and use fc instead.)
	#
	# values that didn't work here in bash 5.0 were:
	#   '', '$()', '$(:)', '$(true)'
	# the value that did was:
	#   '``'
	# This stopped working in bash 5.1, so I guess I was relying on bug
	# behavior. I've had to update this to '`:`' (which now works and
	# seems to be faster than '$(:)'). Note that this behavior change
	# may indicate that the previous emission of what seemed like one
	# sigchld for the subshell and one for anything run inside of it may
	# have itself been a bug (but keep an eye out for this shifting in
	# bash 5.2 again?)
	# shellcheck disable=SC2016,SC2034
	PS0='`:`'
	trap "swain._expand_PS0" SIGCHLD
}

# PROMPT_COMMAND usually runs after some other command,
# but it also gets run to show the initial prompt.
# Let's run a special routine the first time...
PROMPT_COMMAND+=("swain._before_first_prompt")

# The below may have been obsoleted. I've tried running w/o
# it and haven't seen any problems. Dump after 2023
# -H is history expansion (which is turned off by default in scripts)
# this enables fc to work from the script
# set -H


function swain._expand_PS0(){
	# We only want the first SIGCHLD
	# remove the trap so it only fires once
	trap - SIGCHLD

	# noglob enables us to abuse how "$@" works
	# to get each arg quoted to avoid expansion
	# and give the handling hook the choice
	# of whether to force expansion or not
	set -o noglob

	# $1 will be the command number, the command is the rest
	# shellcheck disable=SC2006,SC2046
	# -l lists, -r reverses, -0 is the "newest" item (rest just discards error)
	set -- $(fc -lr -0)
	# preferred $() syntax for above had a bug circa bash 5.0
	# that can mangle user commands if they have more than one
	# command substitution in them; IIRC bash fixed it, but go
	# ahead and retain this knowledge through 2023 in case it
	# bites you.
	set +o noglob

	# shellcheck disable=SC2102,SC2183
	printf -v swain[start_time] '%(%a %b %d %Y %T)T'

	swain[command_number]=$1
	swain[command]="${*:2}" # un-expanded

	# record timestamp once before emitting so that plugins can use it
	# Caution: this might be a misfeature
	swain[start_timestamp]="${EPOCHREALTIME/.}"

	event emit "swain:before_command" "$@"
	# re-record timestamp, which we use to calculate duration, to avoid
	# counting the time taken to run plugin before_command listeners.
	swain[start_timestamp]="${EPOCHREALTIME/.}"
}

trap "event emit 'swain:before_exit'" HUP EXIT

# shellswain has a slightly circuitous init model designed to defer
# work until it's necessary:
# - YOU tell swain WHAT to wrap, and give it a setup callback
# - only if/when the user ever actually invokes the wrapped command, swain will run your callback to fully scaffold any hooks you wanted to apply to the command
# <command> <init callback>
function swain.track(){
	# shellcheck disable=SC2139
	if [[ $(type -ft "$1") == "alias" ]]; then
		echo "shellswain doesn't currently track aliases ($1)"
		return 1
	else
		alias "$1=swain._init_command $1 $2"
		# TODO: There's still a lingering design issue/flaw here; if multiple modules consuming swain try to wrap the same command, there'll be a race condition on who gets to set up. I'm going to put off making a call here because the alias-blocking mechanism should (coincidentally) give the race-loser a clear error. We can cross this bridge if anyone actually wants it. Possibilities:
		# 1. append multiple functions to the same alias if it already exists
		# 2. turn this into an event (I'm trying to avoid the setup costs of this I assume); might be possible to flatten init_command somewhat in this case?
	fi
}

# <command> <init callback>
function swain._init_command(){
	# wire up the normal orchestrator
	# (I don't think swain needs to wire up its *own* before/after callbacks)
	event on "swain:command:$1:run" @_ "swain._run" "$1"

	# re-alias for subsequent runs
	# shellcheck disable=SC2139
	alias "$1=event emit swain:command:$1:run"

	# run <consumer init callback> <command>
	"$2" "$1"

	# give interested parties a chance to init
	# Note: "fire" even runs events added by callbacks
	event fire "swain:command:$1:init"

	# if no hook added a runner, set a default
	if ! event has "swain:phase:run:$1"; then
		# Note: if default runner ever needs to do anything more than just run this, make it a function
		event on "swain:phase:run:$1" @_ command "$1"

	fi

	# kick-off a one-time run of the whole command event cycle
	# (subsequent runs handled by alias)
	event emit "swain:command:$1:run" "${@:3}"
}

# register a callback (and args) for deferred one-time setup
# <command> <callback> [<other args>...]
function swain.hook.init_command(){
	event once "swain:command:$1:init" @_ "$2" "$1" "${@:3}"
}

# <phase> <command> <callback> [<other args>...]
function swain.phase.listen(){
	event on "swain:phase:$1:$2" @_ "$3" "${@:4}"
}

function swain.phase._run(){ # <phase> <command> [<other args>...]
	event emit "swain:phase:${1}:${2}" "${@:3}"
}


function swain._run(){
	swain.phase._run "before" "$@"
	swain.phase._run "run" "$@"
	local ret=$?
	# historical note: at one point this also pulled the record_end again.
	# I think it was a mistake, or perhaps it was working around a bug that I
	# have since fixed? Feel free to delete this after 2023.
	# swain._record_end "${PIPESTATUS[@]}"
	# TODO: I won't do it presumptively, but note that I've thought about
	#       extracting the command-timing bits into a plugin; this staying
	#       dead is a precondition for that. Consider this after June 2023.
	swain.phase._run "after" "$@"
	return $ret
}

# curry some args for a specific phase+command pair
function swain.phase.curry_args(){ # <phase> <command> [<other args>...]
	event curry "swain:phase:$1:$2" "${@:3}"
}
